internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala

```diff
@@ -73,8 +73,50 @@ object ClassToAPI {
                  new mutable.HashSet,
                  new mutable.HashSet)
 
-  def classCanonicalName(c: Class[_]): String =
-    Option(c.getCanonicalName) getOrElse { c.getName }
+  /**
+   * The binary name encoding for Scala has been broken for a while. It does not
+   * follow the Java source name rules because it does not append a `$` after the
+   * mirror class of an object, which already ends in `$`.
+   *
+   * Zinc works around this fix in the only way it can: it gets the enclosing class,
+   * which is obtained from the `InnerClasses` java classfile metadata, and it
+   * reconstructs the path from there. It circumvents, by design, the core issue:
+   * it doesn't call Java's name parser at all.
+   *
+   * This issue has been fixed in the JDK9, but unfortunately it's going to be tormenting
+   * us for a while because 2.13 does not target Java 9, and we're still at 2.12.x.
+   *
+   * <a href="http://bugs.java.com/bugdatabase/view_bug.do?bug_id=8057919">This is the fix in Java 9.</a>
+   *
+   * In order to fully fix this issue, the Scala compiler should do the desired behaviour
+   * in 2.13.x. 2.12.x is already lost since this fix is binary incompatible. The appropriate
+   * JIRA ticket is here: https://github.com/scala/bug/issues/2034.
+   *
+   * @return The canonical name if not null, the full name otherwise.
+   */
+  def handleMalformedNameOf(c: Class[_], isRecursive: Boolean = false): String = {
+    if (c == null) "" // Return nothing if it hits the top-level class
+    else {
+      val className = c.getName
+      // Adds a `dollar` if it's a recursive call or if the original class is an object
+      val atEnd: String =
+        if (isRecursive) "$" else if (className.endsWith("$")) "$" else ""
+      try {
+        val canonicalName = c.getCanonicalName
+        if (canonicalName == null) className
+        else canonicalName + atEnd
+      } catch {
+        case malformedError: java.lang.InternalError
+            if malformedError.getMessage.contains("Malformed class name") =>
+          val enclosingClass = c.getEnclosingClass
+          val enclosingName = enclosingClass.getName
+          val restOfName = c.getName.stripPrefix(enclosingName)
+          handleMalformedNameOf(enclosingClass, true) + restOfName + atEnd
+      }
+    }
+  }
+
+  def classCanonicalName(c: Class[_]): String = handleMalformedNameOf(c)
```

Comment:
Considering this:

A.scala:
```scala
object A {
  def foo = B.C$.x
}
```

B.java:
```java
public class B {
    public static class C$ {
        public static int x = 42;
    }
}
```

The compilation will fail with this patch applied:

```
Caused by: java.lang.RuntimeException: Failed to find name hashes for B.C$
        at scala.sys.package$.error(package.scala:27)
        at sbt.internal.inc.AnalysisCallback.nameHashesForCompanions(Compile.scala:331)
        at sbt.internal.inc.AnalysisCallback.analyzeClass(Compile.scala:338)
        at sbt.internal.inc.AnalysisCallback.$anonfun$addProductsAndDeps$4(Compile.scala:354)
        at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
        at scala.collection.mutable.HashSet.foreach(HashSet.scala:78)
        at scala.collection.TraversableLike.map(TraversableLike.scala:234)
        at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
        at scala.collection.mutable.AbstractSet.scala$collection$SetLike$$super$map(Set.scala:46)
        at scala.collection.SetLike.map(SetLike.scala:101)
        at scala.collection.SetLike.map$(SetLike.scala:101)
        at scala.collection.mutable.AbstractSet.map(Set.scala:46)
        at sbt.internal.inc.AnalysisCallback.$anonfun$addProductsAndDeps$1(Compile.scala:354)
        at scala.collection.TraversableOnce.$anonfun$foldLeft$1(TraversableOnce.scala:157)
        at scala.collection.TraversableOnce.$anonfun$foldLeft$1$adapted(TraversableOnce.scala:157)
        at scala.collection.mutable.HashSet.foreach(HashSet.scala:78)
        at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:157)
        at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:155)
        at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
        at scala.collection.TraversableOnce.$div$colon(TraversableOnce.scala:151)
        at scala.collection.TraversableOnce.$div$colon$(TraversableOnce.scala:151)
        at scala.collection.AbstractTraversable.$div$colon(Traversable.scala:104)
        at sbt.internal.inc.AnalysisCallback.addProductsAndDeps(Compile.scala:350)
        at sbt.internal.inc.AnalysisCallback.get(Compile.scala:300)
        at sbt.internal.inc.Incremental$.doCompile(Incremental.scala:102)
        at sbt.internal.inc.Incremental$.$anonfun$compile$4(Incremental.scala:82)
        at sbt.internal.inc.IncrementalCommon.recompileClasses(IncrementalCommon.scala:118)
        at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:64)
        at sbt.internal.inc.Incremental$.$anonfun$compile$3(Incremental.scala:84)
        at sbt.internal.inc.Incremental$.manageClassfiles(Incremental.scala:129)
        at sbt.internal.inc.Incremental$.compile(Incremental.scala:75)
        at sbt.internal.inc.IncrementalCompile$.apply(Compile.scala:70)
        at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:309)
        at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:267)
        at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:158)
        at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:237)
        at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:68)
        at sbt.internal.inc.ProjectStructure.compile(IncHandler.scala:386)
        at sbt.internal.inc.IncHandler$$anonfun$commands$lzycompute$1.applyOrElse(IncHandler.scala:133)
        at sbt.internal.inc.IncHandler$$anonfun$commands$lzycompute$1.applyOrElse(IncHandler.scala:133)
        at sbt.internal.inc.IncHandler.applyOrElse(IncHandler.scala:197)
        at sbt.internal.inc.IncHandler.$anonfun$noArgs$1(IncHandler.scala:195)
        at sbt.internal.inc.IncHandler.$anonfun$noArgs$1$adapted(IncHandler.scala:195)
        at sbt.internal.inc.IncHandler.$anonfun$apply$1(IncHandler.scala:97)
        at sbt.internal.inc.IncHandler.$anonfun$apply$1$adapted(IncHandler.scala:97)
        at sbt.internal.inc.IncHandler.onIncInstance(IncHandler.scala:104)
        at sbt.internal.inc.IncHandler.apply(IncHandler.scala:98)
        at sbt.internal.inc.IncHandler.apply(IncHandler.scala:52)
        at sbt.internal.scripted.ScriptRunner.processStatement$1(ScriptRunner.scala:18)
        ... 52 more
```